File: scripts/update_sky.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""9bitのSky日替わり情報を取得して data/sky.json を作るスクリプト。
外部ライブラリ不要。GitHub Actionsでそのまま動きます。
"""
from __future__ import annotations
import datetime as dt
import html
import json
import os
import re
import sys
import urllib.request
from html.parser import HTMLParser
from pathlib import Path

URLS = {
    "daily": "https://9-bit.jp/skygold/6593",
    "candle": "https://9-bit.jp/skygold/4920/",
    "shard": "https://9-bit.jp/skygold/23767/",
}
UA = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 Chrome/120 Safari/537.36"
JST = dt.timezone(dt.timedelta(hours=9), "JST")
ROOT = Path(__file__).resolve().parents[1]
OUT = ROOT / "data" / "sky.json"

class TextExtractor(HTMLParser):
    def __init__(self):
        super().__init__()
        self.skip = 0
        self.parts = []
    def handle_starttag(self, tag, attrs):
        if tag in {"script", "style", "noscript", "svg"}:
            self.skip += 1
        if tag in {"p","div","br","li","tr","h1","h2","h3","h4","td","th"}:
            self.parts.append("\n")
    def handle_endtag(self, tag):
        if tag in {"script", "style", "noscript", "svg"} and self.skip:
            self.skip -= 1
        if tag in {"p","div","li","tr","h1","h2","h3","h4"}:
            self.parts.append("\n")
    def handle_data(self, data):
        if not self.skip:
            s = html.unescape(data).strip()
            if s:
                self.parts.append(s)

def fetch(url: str) -> str:
    req = urllib.request.Request(url, headers={"User-Agent": UA, "Accept-Language":"ja,en;q=0.8"})
    with urllib.request.urlopen(req, timeout=25) as r:
        charset = r.headers.get_content_charset() or "utf-8"
        return r.read().decode(charset, errors="replace")

def to_text(src: str) -> str:
    p = TextExtractor(); p.feed(src)
    txt = "\n".join(p.parts)
    txt = re.sub(r"[ \t\u3000]+", " ", txt)
    txt = re.sub(r"\n{2,}", "\n", txt)
    return txt.strip()

def lines_between(text: str, start_pat: str, end_pats: list[str]) -> list[str]:
    lines = [x.strip() for x in text.splitlines() if x.strip()]
    start = 0
    for i,l in enumerate(lines):
        if re.search(start_pat, l):
            start = i + 1
            break
    end = min(len(lines), start + 80)
    for i in range(start, min(len(lines), start + 120)):
        if any(re.search(p, lines[i]) for p in end_pats):
            end = i
            break
    return lines[start:end]

def uniq(seq):
    out=[]
    for x in seq:
        x = re.sub(r"^▲\s*", "", x).strip()
        if x and x not in out:
            out.append(x)
    return out

def parse_daily(text: str) -> dict:
    sec = lines_between(text, r"今日.*デイリークエスト", [r"関連情報", r"デイリークエストとは"])
    target = "取得できませんでした"
    quests=[]
    for l in sec:
        m = re.search(r"主な対象エリア\s*(.+)", l)
        if m: target = m.group(1).strip()
        if re.search(r"一覧", l) or re.search(r"開始時間|終了時間|主な対象エリア", l):
            continue
        # クエスト名っぽい行だけ拾う
        if any(k in l for k in ["呼びかけ", "精霊", "光", "会い", "瞑想", "キャンドル", "プレイヤー", "記憶", "捕まえる", "花", "蟹", "マンタ"]):
            cleaned = re.sub(r"^Image:.*?\s", "", l).strip()
            if 5 <= len(cleaned) <= 70:
                quests.append(cleaned)
    return {"area": target, "quests": uniq(quests)[:4], "source": URLS["daily"]}

def parse_candle(text: str) -> dict:
    sec = lines_between(text, r"今日のデイリー大キャンドル", [r"関連情報", r"日替わり大キャンドルとは"])
    area = "取得できませんでした"
    count = "4個"
    places=[]
    for l in sec:
        m = re.search(r"エリア\s*(.+)", l)
        if m: area = m.group(1).strip()
        m = re.search(r"個数\s*(.+)", l)
        if m: count = m.group(1).strip()
        if l.startswith("▲") or "エリア" in l and ("手前" in l or "岩" in l or "出口" in l or "左" in l or "右" in l or "船" in l or "祠" in l):
            if not re.search(r"対象期間|エリア\s|個数|今日の", l):
                places.append(l)
    return {"area": area, "count": count, "places": uniq(places)[:8], "source": URLS["candle"]}

def parse_shard(text: str) -> dict:
    sec = lines_between(text, r"今日.*闇の破片|本日.*闇の破片|闇の破片.*今日", [r"関連情報", r"闇の破片とは", r"過去"])
    if not sec: sec = text.splitlines()[:120]
    place="取得できませんでした"; reward=""; times=[]; notes=[]
    for l in sec:
        if re.search(r"場所|エリア|落下", l) and not re.search(r"目次|まとめ", l):
            val = re.sub(r"^(場所|エリア|落下場所)\s*", "", l).strip()
            if 2 <= len(val) <= 80: place = val
        if re.search(r"赤|黒|星のキャンドル|通常キャンドル|報酬", l):
            if len(l) <= 90: reward = l
        if re.search(r"\d{1,2}[:：時]\d{0,2}|\d{1,2}時", l):
            if len(l) <= 100: times.append(l)
        if any(k in l for k in ["噴火", "浄化", "火種", "強い", "普通", "ありません"]):
            if len(l) <= 90: notes.append(l)
    return {"place": place, "reward": reward, "times": uniq(times)[:6], "notes": uniq(notes)[:5], "source": URLS["shard"]}

def next_bakery(now: dt.datetime) -> dict:
    # パン焼きは毎時 00/15/30/45 分を目安表示（ユーザー側で補正可能）
    slots=[]
    base = now.replace(second=0, microsecond=0)
    for h in range(0, 24):
        for m in (0,15,30,45):
            t = base.replace(hour=h, minute=m)
            if t <= now: t += dt.timedelta(days=1)
            slots.append(t)
    nxt = min(slots)
    return {"label":"パン焼き", "next": nxt.isoformat(), "memo":"毎時 00 / 15 / 30 / 45 分目安。ゲーム内とズレる場合はアプリの手動補正を使用してください。"}

def main() -> int:
    now = dt.datetime.now(JST)
    result = {"updatedAt": now.isoformat(), "ok": True, "errors": [], "daily": {}, "candle": {}, "shard": {}, "timers": {}}
    for key,url in URLS.items():
        try:
            text = to_text(fetch(url))
            if key == "daily": result[key] = parse_daily(text)
            elif key == "candle": result[key] = parse_candle(text)
            elif key == "shard": result[key] = parse_shard(text)
        except Exception as e:
            result["ok"] = False
            result["errors"].append(f"{key}: {e}")
    result["timers"]["bakery"] = next_bakery(now)
    OUT.parent.mkdir(parents=True, exist_ok=True)
    OUT.write_text(json.dumps(result, ensure_ascii=False, indent=2), encoding="utf-8")
    print(json.dumps(result, ensure_ascii=False, indent=2))
    return 0
if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/update-sky.js
import fs from 'node:fs/promises';
import path from 'node:path';
import * as cheerio from 'cheerio';

const URLS = {
  daily: 'https://9-bit.jp/skygold/6593/',
  candles: 'https://9-bit.jp/skygold/4920/',
  shards: 'https://9-bit.jp/skygold/23767/'
};

const breadTimes = ['01:35','03:35','05:35','07:35','09:35','11:35','13:35','15:35','17:35','19:35','21:35','23:35'];
const geyserTimes = ['00:05','02:05','04:05','06:05','08:05','10:05','12:05','14:05','16:05','18:05','20:05','22:05'];
const turtleTimes = ['00:50','02:50','04:50','06:50','08:50','10:50','12:50','14:50','16:50','18:50','20:50','22:50'];

function jstNow() {
  const d = new Date();
  return new Intl.DateTimeFormat('ja-JP', {
    timeZone: 'Asia/Tokyo', year: 'numeric', month: '2-digit', day: '2-digit',
    hour: '2-digit', minute: '2-digit', second: '2-digit', hour12: false
  }).format(d).replaceAll('/', '-');
}

async function fetchHtml(url) {
  const res = await fetch(url, {
    headers: {
      'User-Agent': 'Mozilla/5.0 (GitHub Actions; Sky tools) AppleWebKit/537.36 Chrome/120 Safari/537.36',
      'Accept-Language': 'ja,en-US;q=0.9,en;q=0.8'
    }
  });
  if (!res.ok) throw new Error(`${url} ${res.status} ${res.statusText}`);
  return await res.text();
}

function linesFromHtml(html) {
  const $ = cheerio.load(html);
  $('script, style, noscript, iframe, form, .comments, #comments').remove();
  const text = $('article').text() || $('body').text();
  return text
    .replace(/\u00a0/g, ' ')
    .split(/\n+/)
    .map(s => s.replace(/\s+/g, ' ').trim())
    .filter(Boolean);
}

function findValue(lines, label) {
  for (let i = 0; i < lines.length; i++) {
    const line = lines[i];
    if (line === label && lines[i + 1]) return lines[i + 1];
    if (line.startsWith(label)) return line.slice(label.length).trim();
  }
  return '';
}

function sliceBetween(lines, startPattern, endPatterns) {
  const start = lines.findIndex(l => startPattern.test(l));
  if (start < 0) return [];
  let end = lines.length;
  for (let i = start + 1; i < lines.length; i++) {
    if (endPatterns.some(p => p.test(lines[i]))) { end = i; break; }
  }
  return lines.slice(start + 1, end);
}

function cleanItem(s) {
  return s
    .replace(/^▲\s*/, '')
    .replace(/^・\s*/, '')
    .replace(/^[0-9０-９]+[.)．、]\s*/, '')
    .trim();
}

function unique(arr) {
  return [...new Set(arr.map(cleanItem).filter(Boolean))];
}

function parseDaily(lines) {
  const area = findValue(lines, '主な対象エリア') || findValue(lines, '対象エリア') || '未取得';
  const start = findValue(lines, '開始時間');
  const end = findValue(lines, '終了時間');
  const section = sliceBetween(lines, /デイリークエスト.*覧/, [/関連情報/, /デイリークエストとは/, /おすすめ記事/]);
  const quests = unique(section.filter(l =>
    !/デイリークエスト|Image|一覧|攻略|開始時間|終了時間|主な対象エリア/.test(l) &&
    l.length >= 4 && l.length <= 80
  )).slice(0, 4);
  return { area, start, end, quests: quests.length ? quests : ['取得できませんでした'] };
}

function parseCandles(lines) {
  const period = findValue(lines, '期間');
  const area = findValue(lines, 'エリア') || '未取得';
  const count = findValue(lines, '個数') || '';
  const section = sliceBetween(lines, /今日のデイリー大キャンドルの場所/, [/関連情報/, /日替わり大キャンドルとは/, /おすすめ記事/]);
  const locations = unique(section.filter(l =>
    (l.includes('▲') || /エリア|広場|神殿|入口|出口|洞窟|塔|祠|船|階|岩|左|右|手前|奥|上|下/.test(l)) &&
    !/Image|今日の|場所$|関連情報|一覧/.test(l) &&
    l.length >= 3 && l.length <= 90
  )).slice(0, 8);
  return { period, area, count, locations: locations.length ? locations : ['取得できませんでした'] };
}

function parseShards(lines) {
  const section = sliceBetween(lines, /今日.*闇の破片|闇の破片.*今日|本日.*闇の破片/, [/関連情報/, /闇の破片とは/, /おすすめ記事/, /コメント/]);
  const target = section.length ? section : lines.slice(0, 120);
  const joined = target.join(' / ');
  const location = findValue(target, '場所') || findValue(target, 'エリア') ||
    (joined.match(/(?:場所|エリア)\s*[:：]?\s*([^/]{2,30})/)?.[1]?.trim() || '未取得');
  const type = joined.includes('赤') ? '赤' : joined.includes('黒') ? '黒' : '未取得';
  const times = unique((joined.match(/\d{1,2}\s*時\s*\d{0,2}\s*分?/g) || [])
    .map(t => t.replace(/\s+/g, '').replace('時', ':').replace(/分$/, '')));
  const notes = unique(target.filter(l => /▲|場所|時間|赤|黒|破片|星のキャンドル|通常キャンドル/.test(l) && l.length <= 100)).slice(0, 6);
  return { location, type, times, notes };
}

async function main() {
  const errors = [];
  let daily = { area: '未取得', quests: [] };
  let bigCandles = { area: '未取得', locations: [] };
  let shards = { location: '未取得', type: '未取得', times: [] };

  try { daily = parseDaily(linesFromHtml(await fetchHtml(URLS.daily))); } catch (e) { errors.push(`daily: ${e.message}`); }
  try { bigCandles = parseCandles(linesFromHtml(await fetchHtml(URLS.candles))); } catch (e) { errors.push(`candles: ${e.message}`); }
  try { shards = parseShards(linesFromHtml(await fetchHtml(URLS.shards))); } catch (e) { errors.push(`shards: ${e.message}`); }

  const ok = daily.quests?.length && !daily.quests[0].includes('取得できません')
    || bigCandles.locations?.length && !bigCandles.locations[0].includes('取得できません')
    || shards.location !== '未取得';

  const data = {
    status: ok ? 'ok' : 'error',
    source: URLS,
    updatedAtJst: jstNow(),
    daily,
    bigCandles,
    shards,
    timers: { bread: breadTimes, geyser: geyserTimes, turtle: turtleTimes },
    errors
  };

  await fs.mkdir('data', { recursive: true });
  await fs.writeFile('data/sky.json', JSON.stringify(data, null, 2), 'utf8');
  await fs.writeFile('data/latest.txt', `${data.updatedAtJst} status=${data.status}\n`, 'utf8');
  console.log(JSON.stringify(data, null, 2));
}

main().catch(async err => {
  await fs.mkdir('data', { recursive: true });
  const data = {
    status: 'error',
    updatedAtJst: jstNow(),
    daily: { area: '未取得', quests: ['取得中にエラーが発生しました'] },
    bigCandles: { area: '未取得', locations: ['取得中にエラーが発生しました'] },
    shards: { location: '未取得', type: '未取得', times: [] },
    timers: { bread: breadTimes, geyser: geyserTimes, turtle: turtleTimes },
    errors: [err.stack || err.message]
  };
  await fs.writeFile('data/sky.json', JSON.stringify(data, null, 2), 'utf8');
  console.error(err);
  process.exit(0);
});
